### ISSUE TITLE
node/peer manager

### ISSUE THREAD

**Issue by kakoc** [freenet/freenet-core#215]
It will be great to have some node manager(as IPFS for example has https://github.com/ipfs/ipfs-desktop). So that a user can control a bandwidth, manage contracts/files, see the world covering(when a location isn't hidden)

**CC** @iduartgomez, @sanity

**Comment by sanity** [comment]
Yes, our current thinking is that a node will expose a web interface that will allow monitoring of stats and the configuration of things like bandwidth limits. Freenet has something similar called "FProxy".

**Comment by sanity** [comment]
🤖 **Relabeled for consistency**

Applied labels:
- `T-feature`
- `P-low`
- `E-medium`
- `A-developer-xp`
- `A-networking`
- `S-needs-design`

**Reasoning:** This is a request for a new node/peer manager UI (new functionality), so T-feature is appropriate. It reads as a nice-to-have user-facing tool rather than urgent or blocking, so P-low. Implementing a desktop/UI manager will take moderate effort (E-medium). The feature spans developer/user tooling, peer/network visibility, and contract/file management, so A-developer-xp, A-networking and A-contracts apply. As a new UI/tool, it needs design discussion before implementation (S-needs-design). Confidence reflects clarity of intent but uncertainty about exact scope and integration complexity.

Previous labels: `E-medium,P-low`

**Comment by sanity** [comment]
# Issue Priority Assessment

Generated: 2026-06-09T14:10:16.083816251+00:00
Issue hash: `acc96822`

## Value Assessment

This issue requests a node manager/dashboard UI for controlling bandwidth, managing contracts, and viewing network topology — similar to IPFS Desktop or the classic Freenet "FProxy" interface.

The issue remains relevant but partially addressed. The codebase has meaningful infrastructure in place: `tray.rs` includes a system tray with an "Open Dashboard" menu item, `config.rs` has `bandwidth_limit`, `total_bandwidth_limit`, and `min_bandwidth_per_connection` fields in `NetworkApiConfig`, and there's a `diagnostics` subsystem and telemetry pipeline (the `freenet service report` command). However, there's no actual web dashboard UI implementation — the tray's "Open Dashboard" action exists as a menu item stub, but there's no HTTP server or frontend serving a monitoring/configuration interface at this time. The `S-needs-design` label is accurate: the surface area (bandwidth controls, contract browsing, ring visualization, peer stats) needs design before implementation.

User impact is moderate but skewed toward non-technical users and node operators. Currently, bandwidth configuration requires editing a config file manually, contract state is only inspectable via CLI (`fdev diagnostics`), and ring topology is invisible without diving into logs or the telemetry dashboard. This creates friction for anyone running a node who isn't a developer — a meaningful barrier for adoption given Freenet's goal of widespread decentralized hosting. Power users can work around this, but casual node operators cannot.

Strategically this aligns with Freenet's roadmap goal of making self-hosted nodes accessible to a broader audience. A management UI would directly advance that goal and differentiate Freenet from infrastructure-only alternatives. However, it's not on the critical path for the core protocol work — the routing, operations state machines, and transport layer have much higher priority. The `P-low` label is appropriate: this is an adoption and UX concern, not a correctness or reliability concern.

There is no blocking relationship here — no other significant PRs or features depend on a management dashboard existing. The urgency is low. The project is still in a phase where core protocol correctness, reliability (the recent `fix(transport)` and `fix(get)` commits), and developer tooling take precedence. A dashboard would be most valuable after the network stabilizes and the focus shifts to end-user adoption.

## Cost Assessment

This is enough context for a cost assessment. Here is my evaluation:

---

**Relevance Check.** The issue is partially addressed by work done since it was filed. The codebase now has a meaningful node web dashboard at `crates/core/src/server/home_page.rs` that shows connection status, peers, subscriptions, operation stats, and transfer metrics rendered as a self-contained HTML page. Bandwidth configuration already exists in `NetworkApiConfig` with `bandwidth_limit`, `total_bandwidth_limit`, and `min_bandwidth_per_connection` fields in `config.rs`. The diagnostics API (`freenet service report`, `NodeDiagnosticsResponse`) provides structured network status. So the issue is not a greenfield ask anymore — a significant portion of the "monitoring" surface is already live.

**Scope and Complexity.** What remains unimplemented is the *interactive management* side: dynamically changing bandwidth limits at runtime without a restart, a UI for managing which contracts are hosted, and geographic visualization ("world covering"). The read-only dashboard already exists; the write path does not. Adding runtime config mutation would require a new HTTP endpoint or WebSocket command path, a mechanism to propagate config changes into the live transport layer (touching `global_bandwidth.rs`, `token_bucket.rs`, and `fixed_rate/controller.rs`), and thread-safe state plumbing from the API layer through to per-connection rate limiters. Contract management (listing, pinning, or evicting hosted contracts) would touch `ring/hosting.rs`, the contract store, and the governance module. Geographic visualization is cosmetic but requires either exposing ring-location data to the dashboard (already partly present via `ring_stats`) or integrating a third-party map library. Total scope is moderate: roughly 4–6 files in core plus new HTTP handler(s), but none of the changes are architecturally isolated.

**Technical Difficulty and Risk.** The bandwidth runtime-change path carries meaningful risk: the transport layer's congestion controllers (BBR, LEDBAT) have their own rate-tracking state and live in async tasks; injecting a new limit mid-session requires either atomic config cells read on each packet or a control-channel message to each connection handler. Getting this wrong could cause bandwidth starvation or instability under load. Contract eviction is also non-trivial since it needs to respect hosting commitments and subscriptions. The web UI additions themselves are low risk (pure HTML generation, no protocol changes), but the backend plumbing for mutation is squarely in the high-risk transport and ring surface areas flagged in `bug-prevention-patterns.md`.

**External Dependencies and Testing Effort.** There are no external service signups or bureaucratic blockers. The `S-needs-design` label is accurate — before implementation, the team needs to decide the API contract for runtime mutation (REST vs. WebSocket command, authentication/authorization model, whether bandwidth changes apply globally or per-peer). Testing the monitoring-only parts is easy via unit tests against `NetworkStatus` snapshots. Testing runtime bandwidth mutation is harder: it requires either DST simulation harness changes or manual multi-node tests to verify the rate limiting actually takes effect under real traffic, since the transport-layer tests today do not exercise config hot-reload. Given that the most valuable new capability (runtime bandwidth control) is also the riskiest and hardest to test, the effective implementation cost is higher than the `E-medium` label suggests for a complete, safe implementation.

---
*Generated by issue-prioritizer*

**Comment by sanity** [comment]
🤖 **Automated ranking by Torvan.** Not a maintainer's assessment of this issue.

| Measure | Score |
|---|---:|
| Value | 0.049 |
| Cost | 0.887 |
| Priority | 583 of 593 |

Value and cost run 0 to 1 across the 593 open issues in the freenet repositories: value 1.00 is the most valuable, cost 0.00 the cheapest. Issues are compared against each other rather than scored in isolation. The queue alternates between the most valuable work and the best value for the effort, so neither expensive-but-important nor cheap-and-quick work crowds the other out.

Ranked 2026-08-31, and liable to change. Adding information here, whether that is scope, a reproduction, or a correction to this assessment, causes the issue to be reassessed and this comment updated.
